src/common/AdpfWrapper.cpp

```diff
@@ -0,0 +1,123 @@
+typedef void (*APH_reportActualWorkDuration)(APerformanceHintSession*, int64_t);
+typedef void (*APH_closeSession)(APerformanceHintSession* session);
+
+static bool gAPerformanceHintBindingInitialized = false;
+static APH_getManager gAPH_getManagerFn = nullptr;
+static APH_createSession gAPH_createSessionFn = nullptr;
+static APH_reportActualWorkDuration gAPH_reportActualWorkDurationFn = nullptr;
+static APH_closeSession gAPH_closeSessionFn = nullptr;
+
+static int loadAphFunctions() {
+    if (gAPerformanceHintBindingInitialized) return true;
+
+    void* handle_ = dlopen("libandroid.so", RTLD_NOW | RTLD_NODELETE);
+    if (handle_ == nullptr) {
+        return -1000;
+    }
+
+    gAPH_getManagerFn = (APH_getManager)dlsym(handle_, "APerformanceHint_getManager");
+    if (gAPH_getManagerFn == nullptr) {
+        return -1001;
+    }
+
+    gAPH_createSessionFn = (APH_createSession)dlsym(handle_, "APerformanceHint_createSession");
+    if (gAPH_getManagerFn == nullptr) {
+        return -1002;
+    }
+
+    gAPH_reportActualWorkDurationFn = (APH_reportActualWorkDuration)dlsym(
+            handle_, "APerformanceHint_reportActualWorkDuration");
+    if (gAPH_getManagerFn == nullptr) {
+        return -1003;
+    }
+
+    gAPH_closeSessionFn = (APH_closeSession)dlsym(handle_, "APerformanceHint_closeSession");
+    if (gAPH_getManagerFn == nullptr) {
+        return -1004;
+    }
+
+    gAPerformanceHintBindingInitialized = true;
+    return 0;
+}
+
+bool AdpfWrapper::sUseAlternativeHack = false; // FIXME remove hack
+
+int AdpfWrapper::open(pid_t threadId,
+                      int64_t targetDurationNanos) {
+    std::lock_guard<std::mutex> lock(mLock);
+    int result = loadAphFunctions();
+    if (result < 0) return result;
+
+    // This is a singleton.
+    APerformanceHintManager* manager = gAPH_getManagerFn();
+
+    int32_t thread32 = threadId;
+    if (sUseAlternativeHack) {
+        // FIXME hack because setprop not working!
+        targetDurationNanos = (targetDurationNanos & ~0xFF) | 0xA5;
+    }
+    mHintSession = gAPH_createSessionFn(manager, &thread32, 1 /* size */, targetDurationNanos);
+    if (mHintSession == nullptr) {
+        return -1;
+    }
+    return 0;
+}
+
+void AdpfWrapper::reportActualDuration(int64_t actualDurationNanos) {
+    //LOGD("ADPF Oboe %s(dur=%lld)", __func__, (long long)actualDurationNanos);
+    std::lock_guard<std::mutex> lock(mLock);
+    if (mHintSession != nullptr) {
+        gAPH_reportActualWorkDurationFn(mHintSession, actualDurationNanos);
+    }
+}
+
+void AdpfWrapper::close() {
+    std::lock_guard<std::mutex> lock(mLock);
+    if (mHintSession != nullptr) {
+        gAPH_closeSessionFn(mHintSession);
+        mHintSession = nullptr;
+    }
+}
+
+void AdpfWrapper::onBeginCallback() {
+    if (isOpen()) {
+        mBeginCallbackNanos = oboe::AudioClock::getNanoseconds(CLOCK_REALTIME);
+    }
+}
+
+void AdpfWrapper::onEndCallback(double durationScaler) {
```

Comment:
This seems to be only called when durationScaler < 2.0. Is this expected?

Reply:
It is OK to skip this if we have a short callback to reduce error. ADPF still works on sparse data.